src/cljc/quil/middlewares/navigation_2d.cljc

```diff
@@ -13,24 +13,21 @@
     (throw #?(:clj (RuntimeException. missing-navigation-key-error)
               :cljs (js/Error. missing-navigation-key-error)))))
 
-(defn- default-position
-  "Default position configuration: zoom is neutral and central point is
-  `width/2, height/2`."
-  []
-  {:position [(/ (q/width) 2.0)
+(defn- default-settings []
+  {:width (q/width)
+   :height (q/height)
```

Comment:
The only reason was to make `screen->world-coords` and `world->screen-coords` pure functions rather than depending on some external state. I thought that would make navigation-2d more flexible, for instance allowing the user to apply it to some smaller rectangle inside their window instead of to the whole window. But you're right that it would require the user to update the width and height in the state map if the window resized, which seems annoying and redundant. Also, I guess the caching could break someone's code if they're already using navigation-2d with window resizing.

However, I've just realized that the only reason navigation-2d needs to know the dimensions of the window in the first place is because, when you zoom, it zooms in on the center of the window. If it was changed so that it zooms in on the mouse (which is what google maps does for example), then we wouldn't need to know the width and height at all.

Even if we leave the zoom centered on the center of the window, `screen->world-coords` and `world->screen-coords` could still be rewritten to not need the width and the height (and simplifying them a lot in the process). That way all the publicly exposed functions would be pure, and only the private event handlers would need to access the window's dimensions. In order to make this rewrite possible, the state map would have to store the world coordinates of the top left corner of the window (the `:position` key that's already there stores the world coordinates of the center of the window). I'd hesitate to change the `:position` key to point to the top left instead of the center since that could theoretically break someone's code if they were manually messing with `:position`. On the other hand, it's redundant to store both the top left corner and the center and they could become temporarily inconsistent if someone's code is updating `:position` manually (but no old code would be broken).

Perhaps the best thing to do is to put my changes in a separate middleware as you suggested, but I don't know what I would name it haha.